### PR TITLE
Refactor Char and String generators

### PR DESCRIPTION
**Pull request by ashawley**
Code cherry-picked from https://github.com/typelevel/scalacheck/pull/575

### CODE REVIEW

**Comment by ashawley** [comment]
This is a refactor that does make considerable performance improvements.  The string tests are about 10x faster.

```
> bench/jmh:run -wi 3 -i 3 -f1 -t1 org.scalacheck.bench.GenBench
Benchmark                   (genSize)  (seedCount)  Mode  Cnt      Score       Error  Units
GenBench.arbitraryString          100          100  avgt    3    397.990 ±   208.168  us/op
GenBench.asciiPrintableStr        100          100  avgt    3    206.355 ±   115.595  us/op
GenBench.const_                   100          100  avgt    3      4.404 ±     2.266  us/op
GenBench.double_                  100          100  avgt    3     13.328 ±     7.818  us/op
GenBench.dynamicFrequency         100          100  avgt    3   2089.066 ±  1196.049  us/op
GenBench.eitherIntInt             100          100  avgt    3     58.887 ±    36.921  us/op
GenBench.identifier               100          100  avgt    3    308.474 ±   191.908  us/op
GenBench.int_                     100          100  avgt    3     17.041 ±    10.307  us/op
GenBench.listOfInt                100          100  avgt    3   2058.417 ±  2079.534  us/op
GenBench.mapOfIntInt              100          100  avgt    3   4675.506 ±  3840.344  us/op
GenBench.oneOf                    100          100  avgt    3     25.371 ±    10.698  us/op
GenBench.optionInt                100          100  avgt    3     80.699 ±    57.109  us/op
GenBench.sequence                 100          100  avgt    3    304.381 ±   191.811  us/op
GenBench.staticFrequency          100          100  avgt    3   2018.222 ±  1040.656  us/op
GenBench.testFilter               100          100  avgt    3  24008.234 ± 30648.807  us/op
GenBench.zipIntInt                100          100  avgt    3     47.855 ±    60.774  us/op
```

**Comment by ashawley** [comment]
The Travis build still takes 58min total time.  That may be because there aren't a lot of string or character tests in the test suite.

File: src/main/scala/org/scalacheck/Gen.scala [hunk 1]
```diff
@@ -790,86 +790,120 @@ object Gen extends GenArities with GenVersionSpecific {
...
+        val (x, seed1) = seed0.long
+        val i = ((x & Long.MaxValue) % cs.length).toInt
+        r(Some(cs(i)), seed1)
+      }
+    }
+
   /** Generates a numerical character */
-  val numChar: Gen[Char] =
-    choose(48.toChar, 57.toChar)
+  def numChar: Gen[Char] =
```
Comment (ashawley) [inline src/main/scala/org/scalacheck/Gen.scala]: Whoops this reverted the def to val change.

**Comment by ashawley** [comment]
I cherry-picked an intermediate commit, https://github.com/typelevel/scalacheck/pull/575/commits/3cbcf07b700a69b1aec7f7573165c21e727b9515, here as 866ce69, but the benchmark indicates it slowed things down.  I'll cherry-pick the third commit and see where the benchmarks land.

**Comment by ashawley** [comment]
It does seem that `arbitraryString` does appear to jump up, most likely because of the modifications to `Gen.mkString`.  Strangely, it doesn't affect `asciiPrintableStr` or `identifier` benchmarks.  I would have thought that it would have.

```
> bench/jmh:run -wi 3 -i 3 -f 1 -t 1 -w 2 -r 2
Benchmark                   (genSize)  (seedCount)  Mode  Cnt      Score      Error  Units
GenBench.arbitraryString          100          100  avgt    3    852.027 ±  591.165  us/op
GenBench.asciiPrintableStr        100          100  avgt    3    274.318 ±   58.040  us/op
GenBench.const_                   100          100  avgt    3      4.530 ±    0.548  us/op
GenBench.double_                  100          100  avgt    3     13.527 ±    4.367  us/op
GenBench.dynamicFrequency         100          100  avgt    3   2203.155 ±  309.073  us/op
GenBench.eitherIntInt             100          100  avgt    3     59.916 ±   18.522  us/op
GenBench.identifier               100          100  avgt    3    283.954 ±  138.022  us/op
GenBench.int_                     100          100  avgt    3     17.202 ±    1.531  us/op
GenBench.listOfInt                100          100  avgt    3   2110.610 ± 1298.623  us/op
GenBench.mapOfIntInt              100          100  avgt    3   4773.228 ± 1169.756  us/op
GenBench.oneOf                    100          100  avgt    3     26.354 ±    6.113  us/op
GenBench.optionInt                100          100  avgt    3     72.908 ±    6.498  us/op
GenBench.sequence                 100          100  avgt    3    305.679 ±   12.076  us/op
GenBench.staticFrequency          100          100  avgt    3   2144.414 ±  591.506  us/op
GenBench.testFilter               100          100  avgt    3  24645.296 ± 8130.122  us/op
GenBench.zipIntInt                100          100  avgt    3     45.386 ±    4.124  us/op
```

**Comment by ashawley** [comment]
> It does seem that `arbitraryString` does appear to jump up, most likely because of the modifications to `Gen.mkString`.

I was wrong.  Digging in to this further, it appears that this small change causes  the benchmark`arbitraryString` to double:

```patch
--- b/src/main/scala/org/scalacheck/Gen.scala
+++ a/src/main/scala/org/scalacheck/Gen.scala
@@ -1056,7 +1056,7 @@
   private def mkSize(p: Gen.Parameters, seed0: Seed): (Int, Seed) = {
     val maxSize = Integer.max(p.size + 1, 1)
     val (x, seed1) = seed0.long
-    ((x % maxSize).toInt, seed1)
+    (((x & Long.MaxValue) % maxSize).toInt, seed1)
   }
 
   // used to calculate how many per-item retries we should allow.
```

**Comment by non** [comment]
@ashawley What's the status here? Happy to merge this as-is and keep plugging away -- I think your approach of chipping away at my massive refactor probably makes sense.

One note: the `arbitraryString` benchmark doesn't appear in this branch. How were you running it? I'm happy to revert the change you found that makes it slower (the uniform size calculation) although I don't fully understand why that would have that effect.

**Comment by ashawley** [comment]
Sorry, that I didn't finish this off.  Yes, that's true the benchmark is missing from the branch.  I would just keep the benchmark modified in my working git tree while I worked on each of each of these separate branches.

**Comment by ashawley** [comment]
Yes, it doesn't make sense to me either why that small change causes a performance hit.

I'll work on updating the branch since it's conflicted, so it's easier to study this further by including bringing in the `arbitraryString` benchmark.

**Comment by ashawley** [comment]
Actually, I'll close this and reopen since the commit hashes are referenced.

**Pull request closed without merging**